File: sample/SampleObjC/CustomScanOverlayViewController.m
//
//  CustomScanOverlayViewController.m
//
//  Copyright © 2019 Jumio Corporation All rights reserved.
//

#import "CustomScanOverlayViewController.h"

@interface CustomScanOverlayViewController ()

@property (nonatomic, strong) IBOutlet UIButton *flashButton;
@property (nonatomic, strong) IBOutlet UIButton *cameraButton;
@property (weak, nonatomic) IBOutlet UIButton *closeButton;
@property (weak, nonatomic) IBOutlet UIButton *startStopButton;

@property (nonatomic, assign) BOOL isScanning;

@end

@implementation CustomScanOverlayViewController

- (instancetype) initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        //To change the vertical card frame offset in portrait orientation, set the following property (default = 0).
        self.verticalCardFrameOffset = -100.0;
        
        self.delegate = self;
        self.isScanning = NO;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.translatesAutoresizingMaskIntoConstraints = NO;
    
    _flashButton.layer.cornerRadius = 20.0;
    _cameraButton.layer.cornerRadius = 20.0;
    _closeButton.layer.cornerRadius = 20.0;
    _startStopButton.layer.cornerRadius = 20.0;
}

//Once viewWillAppear is called within your BAMCheckoutCustomScanOverlayViewController, you can perform the following actions:
// - Get location and dimension of card frame
// - Check if front and back camera available
// - Get camera position (front or back)
// - Switch between front and back camera
// - Check if flash available
// - Check if flash enabled
// - Switch flash mode (on or off)
// - Restart card scanning if retry possible
// - Stop card scanning

- (void) viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    if (![self hasFlash]) _flashButton.hidden = YES;
    if (![self hasMultipleCameras]) _cameraButton.hidden = YES;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (IBAction)flashBtnTapped:(id)sender {
    [self toggleFlash];
    
    NSString *btnTitle = ([self isFlashOn]) ? @"F+" : @"F-";
    [_flashButton setTitle:btnTitle forState:UIControlStateNormal];
}

- (IBAction)cameraBtnTapped:(id)sender {
    [self switchCamera];
    
    NSString *btnTitle = ([self currentCameraPosition] == JumioCameraPositionBack) ? @"C^" : @"Cv";
    [_cameraButton setTitle:btnTitle forState:UIControlStateNormal];
    
    if ([self hasFlash]) {
        _flashButton.hidden = NO;
    } else {
        _flashButton.hidden = YES;
    }
}

- (IBAction)close:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)startStop:(id)sender {
    if (self.isScanning) {
        [self stopScan];
    } else {
        [self retryScan];
    }
}

- (void) retryScan {
    [super retryScan];
    
    self.isScanning = YES;
    [self updateCaptureButton];
}

- (void) stopScan {
    [super stopScan];
    
    self.isScanning = NO;
    [self updateCaptureButton];
}

- (void) updateCaptureButton {
    if (self.isScanning) {
        [_startStopButton setTitle:@">" forState:UIControlStateNormal];
    } else {
        [_startStopButton setTitle:@"||" forState:UIControlStateNormal];
    }
}


#pragma mark - BAMCheckoutCustomScanOverlayViewControllerDelegate

/**
 * This method is called when an error occured while the ScanViewController is opend
 * @param error The error codes 200, 210, 220, 240, 300 and 320 will be returned.
 * @param retryPossible indicates whether [self retryScan] can be called or not
 **/
- (void) bamCheckoutStoppedWithError:(NSError *)error retryPossible:(BOOL)retryPossible {
    NSString *retryString = (retryPossible) ? @"retry possible" : @"no retry possible";

    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:[NSString stringWithFormat:@"%ld", (long)error.code]
                                                                             message:error.localizedDescription
                                                                      preferredStyle:UIAlertControllerStyleAlert];
    
     [alertController addAction:[UIAlertAction actionWithTitle:@"close" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
         [self dismissViewControllerAnimated:YES completion:nil];
    }]];
    
    [alertController addAction:[UIAlertAction actionWithTitle:retryString style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self retryScan];
    }]];

    
     [self presentViewController:alertController animated:YES completion:nil];

}

@end

File: sample/SampleSwift/CustomScanOverlayViewController.swift
//
//  CustomScanOverlayViewController.swift
//
//  Copyright © 2019 Jumio Corporation All rights reserved.
//

import BAMCheckout

class CustomScanOverlayViewController : BAMCheckoutCustomScanOverlayViewController, BAMCheckoutCustomScanOverlayViewControllerDelegate {
    
    @IBOutlet weak var flashButton:UIButton!;
    @IBOutlet weak var cameraButton:UIButton!;
    @IBOutlet weak var closeButton:UIButton!;
    @IBOutlet weak var startStopButton:UIButton!;
    
    var isScanning:Bool;
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        
        self.isScanning = false
        
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        
        //To change the vertical card frame offset in portrait orientation, set the following property (default = 0).
        self.verticalCardFrameOffset = -100.0
        
        self.delegate = self
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.translatesAutoresizingMaskIntoConstraints = false
        
        self.flashButton.layer.cornerRadius = 20.0
        self.cameraButton.layer.cornerRadius = 20.0
        self.closeButton.layer.cornerRadius = 20.0
        self.startStopButton.layer.cornerRadius = 20.0
    }
    
    @IBAction func flashBtnTapped() -> Void {
        self.toggleFlash()
        
        let btnTitle:String = self.isFlashOn() ? "F+" : "F-"
        self.flashButton.setTitle(btnTitle, for: UIControlState.normal)
    }
    
    @IBAction func cameraBtnTapped() -> Void {
        self.switchCamera()
        
        let btnTitle:String = (self.currentCameraPosition() == JumioCameraPositionBack) ? "C^" : "Cv"
        self.cameraButton.setTitle(btnTitle, for: UIControlState.normal)
        
        if (self.hasFlash()) {
            self.flashButton.isHidden = false
        } else {
            self.flashButton.isHidden = true
        }
    }
    
    @IBAction func close() ->Void {
        self.dismiss(animated: true, completion: nil)
    }
    
    @IBAction func startStop() -> Void {
        if (self.isScanning) {
            self.stopScan()
        } else {
            self.retryScan()
        }
    }
    
    override func retryScan() {
        super.retryScan()
        
        self.isScanning = true
        self.updateCaptureButton()
    }
    
    override func stopScan() {
        super.stopScan()
        
        self.isScanning = false
        self.updateCaptureButton()
    }
    
    func updateCaptureButton() -> Void {
        if (self.isScanning) {
            self.startStopButton.setTitle(">", for: UIControlState.normal)
        } else {
            self.startStopButton.setTitle("||", for: UIControlState.normal)
        }
    }
    
    func bamCheckoutStoppedWithError(_ error: Error!, retryPossible: Bool) {
        
        let retryString = retryPossible ? "retry possible" : "no retry possible"
        
        let alertController:UIAlertController = UIAlertController.init(title: String(describing: error), message: error.localizedDescription, preferredStyle: UIAlertControllerStyle.alert)
        
        alertController.addAction(UIAlertAction.init(title: retryString, style: UIAlertActionStyle.cancel, handler: { (action: UIAlertAction!) in
            self.dismiss(animated: true, completion: nil)
        }))
        
        self.present(alertController, animated: true, completion: nil)
    }
}
